File: Stack/Daily_Temperatures/2_Better.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        int n = T.size();
        vector<int> ans(n, 0);
        for (int i = n - 2; i >= 0; --i) {
            int j = i + 1;
            while (j < n && T[j] <= T[i]) {
                if (ans[j] == 0) { // no warmer day ahead of j
                    j = n; // break outer while
                    break;
                }
                j += ans[j]; // jump to the next candidate
            }
            if (j < n) ans[i] = j - i;
        }
        return ans;
    }
};

File: Stack/Daily_Temperatures/2_Better.java
import java.util.*;
class Solution {
    public int[] dailyTemperatures(int[] T) {
        int n = T.length;
        int[] ans = new int[n];
        for (int i = n - 2; i >= 0; i--) {
            int j = i + 1;
            while (j < n && T[j] <= T[i]) {
                if (ans[j] == 0) { // no warmer day after j
                    j = n; // break loop
                    break;
                }
                j += ans[j]; // jump ahead
            }
            if (j < n) ans[i] = j - i;
        }
        return ans;
    }
}
